Added OpenStack Swift integration with CaffeOnSpark Documentation

arundasan91 opened the pull request:
Adding OpenStack Swift integration with CaffeOnSpark Documentation and few scripts to aid in getting the correct version of hadoop, spark. Added an editted `pom.xml` file to account for `hadoop-openstack` dependency while rebuilding CaffeOnSpark.
Please take a look.

https://github.com/arundasan91/CaffeOnSpark/wiki/GetStarted_yarn_object_storage

yahoocla commented:
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

yahoocla commented:
CLA is valid!

anfeng commented:
@arundasan91 I would rather don't create yet another pom.xml. Comparing with the existing pom.xml, I see 2 differences:
   * spark.version: 2.0.0 --> 2.0.1
      * Is it required to use 2.0.1?
   * <dependency>
         <groupId>org.apache.hadoop</groupId>
         <artifactId>hadoop-openstack</artifactId>
         <version>2.7.1</version>
       </dependency>

Sounds like that we just need to include that in documentation, not a new pom.xml

arundasan91 commented:
I used the latest version of Spark v2.0.1.

The version of Hadoop is what matters(v2.7.1). Spark can be v2.0.0.

I get you. We could just add that in the documentation.

anfeng commented on scripts/scripts_object_storage/openstack_swift/local-setup-spark.sh:
```diff
@@ -0,0 +1,10 @@
+#
+# This script install Spark locally
+
+wget http://archive.apache.org/dist/spark/spark-2.0.1/spark-2.0.1-bin-hadoop2.7.tgz
```
Can use http://archive.apache.org/dist/spark/spark-2.0.0/spark-2.0.0-bin-hadoop2.7.tgz instead? It will reduce the changes.

arundasan91 replied:
Yes, we definitely can use http://archive.apache.org/dist/spark/spark-2.0.0/spark-2.0.0-bin-hadoop2.7.tgz.

"Why do we need hadoop-openstack as a dependency into pom.xml? In our pomx.xml, Apache Spark etc are included for compilation purpose only, and thus scope "provided". hadoop-openstack should be given in your Hadoop cluster, not in CaffeOnSpark."
  ~ I refered https://spark.apache.org/docs/1.5.1/storage-openstack-swift.html and came to a conclusion that `hadoop-openstack` dependency should be added to CaffeOnSpark as well.

Now that you asked the question, I kind of question my approach as well. 

The most important step was to get `hadoop-openstack-2.7.1.jar` file ready in `HADOOP_CLASSPATH`. Do you want me to try this again without modifying the CaffeOnSpark/caffe-grid/pom.xml file with any hadoop-openstack dependency ?

Thanks,
Arun

anfeng replied:
I see your motivation for hadoop-openstack.jar. Placing them into pom.xml will make user's life easier. Otherwise, they will need to download the jar, and add spark-submit argument '-jars'.

Let's introduce a profile specific dependency? We can name the profile as openstack or swift, and that profile will introduce hadoop-openstack as a new dependency.

arundasan91 replied:
Yes, exactly. I spent a few minutes trying to figure out which pom.xml file to target and where to put those in.

About profiles, to be frank, I am not well versed with that. Could you please elaborate ? I'm still in the learning phase.

Thanks,
Arun

anfeng replied:
object_storage is fine.

anfeng commented:
In the GetStarted guide, can you make it easy for a user to understand the changes from the standard YARN guide.

arundasan91 commented:
Sure. Shall I remove the portion of pom.xml also ?

arundasan91 commented:
I updated the Wiki. Please let me know if that helped a bit.
Thanks.

arundasan91 commented:
Thanks for all the help. I learned a lot during this week !

Please take a look at the wiki at https://github.com/arundasan91/CaffeOnSpark/wiki/GetStarted_yarn_object_storage. I updated the `make build object_storage=1` section.

Thanks,
Arun

anfeng commented:
Are you sure this haoop-openstack module is required for S3 access? The [documentation](https://www.cloudera.com/documentation/enterprise/5-5-x/topics/spark_s3.html) from Cloudera indicates that we only need to modify core-site.xml.  

Please simplify the introduction section to state: 
   * This new guide is modified from the GetStarted_yarn so that data sources stored in object storage nodes such as OpenStack Swift could be directly accessed. 
   * Data sources in OpenStack Swift is accessed via swift://container.PROVIDER/path. See https://spark.apache.org/docs/1.5.1/storage-openstack-swift.html as a reference.

It will be nice to mark which section is different from the standard yarn guide.

arundasan91 commented:
hadoop-openstack module is required only for OpenStack Swift access. The directory `$HADOOP_HOME/share/hadoop/tools/lib` which hosts most of the `jar` files, have `hadoop-aws-2.7.1.jar` and `hadoop-openstack-2.7.1.jar` already in them. We need to have these files in Hadoop classpath.
I have ensured that in Section II using:
`export HADOOP_CLASSPATH=${HADOOP_HOME}/share/hadoop/tools/lib/*`

Let me go ahead an verify a few things on a new instance. First I will try building CaffeOnSpark without the `hadoop-openstack` dependency and will see if I can still get Swift to work with it without specifying any extra jar files during training. 

I will update back in some time.

arundasan91 commented:
@anfeng , In the process of trying out different things, I found a bug. Could you please confirm this:

In the Makefile, Line 9, 
```
${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | awk '{print $$5}' | cut -d'.' -f1
```
should be changed to:
```
${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | awk '{print $5}' | cut -d'.' -f1
```

The spark version (v2) is not read properly.

anfeng commented:
@arundasan91 Good finding. Please submit a separated PR for Spark 2.

arundasan91 commented:
Yes. When I change the `SPARK_HOME` in Makefile to my spark installation, I see `mvn -Dspark2 clean` in the log when running `make clean`.

arundasan91 commented:
In another note, I just clarified something that was bugging me for a few hours and this is major.

There is no necessity to add `hadoop-openstack` to pom.xml. Only `core-site.xml` file has to be changed. All the other change should be done only for `hadoop` and `spark`.

I confirmed it with two fresh installs.

Shall I remove the changes in pom.xml ?

Thanks.

arundasan91 commented:
This means I should revert the changes added in `Makefile` and `pom.xml`. This was rather a new experience for me. Sorry for the confusion. My intension was to add openstack dependencies to CaffeOnSpark so that we can use it without specifying other jar files.

I think the only problem was in the Makefile #167.
Also, changes won't matter without the correct path to SPARK_HOME in Makefile.

anfeng commented:
You should also update document accordingly. Maybe you could have it only contain swift specific secstions, and refer other sections on GetStarted_yarn.

arundasan91 commented:
Sure. I will make the changes and add just the Swift specific section as you mentioned.

Thanks for the help.

arundasan91 commented:
Sorry, I was busy with some work. I will update the documentation in some days.

arundasan91 commented:
I accidently removed this patch, patch-1 while cleaning up the repo. Re-opening.

anfeng commented:
Please work on the GetStarted guide per our discussion. Once we have that done, we will close this PR.

arundasan91 commented:
I will work on it tomorrow. 

Did you check Line 13 of the Makefile:
`export SPARK_VERSION=$(shell ${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | awk '{print $$5}' | cut -d'.' -f1)`
It does not work. We need to change it to:
`export SPARK_VERSION=$(shell ${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | cut -d' ' -f10 | cut -d'.' -f1)``

arundasan91 commented:
Could you please take a look at the wiki `https://github.com/arundasan91/CaffeOnSpark/wiki/GetStarted_yarn_object_storage` when you have time ?
Thanks.

anfeng commented:
@arundasan91 The current version of GetStarted is very confusing. Please simplify it to highlight the difference, and have it refer to existing guides whenever feasible. This guide could be very short.

arundasan91 commented:
@anfeng Sorry about that. I've redone the entire wiki. I can make other changes if any after a few hours.

Thanks,
Arun

anfeng commented:
I like the revised guide better. A few additional comments:

- Please make the Introduction section more readable.  
- Let's openstack_swift/core-site.xml  to openstack_swift/core-site.xml.template. In the corresponding guide section, please mention that user provide ALL VALUES for placeholders, such as AUTH URL, USERNAME and REGION.

arundasan91 commented:
I've made some changes to the wiki sir. Added section for `core-site.xml.template`. Users will now edit the template and rename it afterwards. The updated, renamed `core-site.xml` file will be then copied to `$SPARK_HOME/conf`. Please take a look. I asked a few colleagues to go through the introduction and got a green flag. Still, could you please give me some pointers to further improve it ?

anfeng commented:
Let's avoid mentioning Amazon S3 in this version of GetStarted guide. since we have not figured the details for S3 yet. We could always them in later.

Your Introduction section could be simplified a little further. Reference to others docs could be moved to Section I.

arundasan91 commented:
Okay, I will work on it now.

arundasan91 commented:
I have updated the wiki. Please take a look. I will make any other necessary changes to finish the wiki tomorrow.


I could have tested S3 easily if I had access to AWS. I will ask around. 
Thanks,
Arun

anfeng commented:
Please push the guide to CaffeOnSpark.wiki. Great work!

anfeng commented:
+1

arundasan91 commented:
I do not have access to push the wiki right ? I just tried to push the changes to CaffeOnSpark.wiki and failed due to restricted access. My understanding is that Wiki pages (unlike regular repos) can only be merged by owners of the project given the wiki link. Am I missing something ?
Sorry, I am new to this.

arundasan91 commented:
@anfeng , in the [Makefile line 13](https://github.com/yahoo/CaffeOnSpark/blob/master/Makefile#L13), spark version isn't fetched properly now. Could you please look into this ?

This is what I get now:

```
${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | awk '{print $$5}' | cut -d'.' -f1

gives:
/___/
```
This works:

```
${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | cut -d' ' -f10

gives:
1.6.0
```

anfeng commented:
${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | awk
'{print $5}' | cut -d'.' -f1

returns

1


I believe that extra $ for $$5 was added for Makefile


Andy

On Tue, Jan 17, 2017 at 5:29 PM, Arun Das <notifications@github.com> wrote:

> @anfeng <https://github.com/anfeng> , in the Makefile line 13
> <https://github.com/yahoo/CaffeOnSpark/blob/master/Makefile#L13>, spark
> version isn't fetched properly now. Could you please look into this ?
>
> This is what I get now:
>
> ${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | awk '{print $$5}' | cut -d'.' -f1
>
> gives:
> /___/
>
> This works:
>
> ${SPARK_HOME}/bin/spark-submit --version 2>&1 | grep version | cut -d' ' -f10
>
> gives:
> 1.6.0
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/yahoo/CaffeOnSpark/pull/162#issuecomment-273355553>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AClTeJ-wfUXAwl73ps3MdlNTQflb33UAks5rTWrtgaJpZM4KSKyd>
> .
>

arundasan91 commented:
That is correct.

arundasan91 commented:
@anfeng , as I just [tested and commented](https://github.com/yahoo/CaffeOnSpark/pull/224#issuecomment-273922511) in #224 , the extra $ is intensional as you [mentioned](https://github.com/yahoo/CaffeOnSpark/pull/224#issuecomment-273908969).

Thanks for looking into this. I forgot to take into account the way shell scripts specified within a Makefile work.
